### ISSUE TITLE
VS 2022 Xamarin Debugging is not reliable: "Project is not deployable"

### ISSUE THREAD

**Issue by dsaf** [xamarin/xamarin-android#6975]
### Android application type

Classic Xamarin.Android (MonoAndroid12.0, etc.)

### Affected platform version

VS 2022 17.1.6

### Description

Frequently cannot debug Android projects. It most definitely works every now and then, so my Android device is fine etc.

![image](https://user-images.githubusercontent.com/3845650/166224848-c848361b-0aaa-4ad4-9ead-90f89413f58e.png)

![image](https://user-images.githubusercontent.com/3845650/166224877-dec8c349-a579-4341-a282-c075d20d58bb.png)

![image](https://user-images.githubusercontent.com/3845650/166224931-e1647eb3-1ce1-41b3-8ddb-27fd91ee8a8c.png)

![image](https://user-images.githubusercontent.com/3845650/166224743-772d3bfe-15c1-4899-8f3c-e205095f8000.png)

![image](https://user-images.githubusercontent.com/3845650/166227282-cc76844d-fd62-439c-932f-a283f41fe04c.png)

![image](https://user-images.githubusercontent.com/3845650/166224801-7136afe7-74f3-40f3-8f26-ed520a46b3ef.png)

### Steps to Reproduce

1. Use Xamarin.
2. Debugging experience is not reliable.

### Relevant log output

```shell
Project is not deployable.
Project is not deployable.
Project is not deployable.
```

**Comment by dsaf** [comment]
It was never particularly reliable in 2019, but the 2022 seems to be even worse. Would upgrading to .NET 6 make things any better? I could not find a project template for that for some reason. Is it also the 80s-style terminal experience using `dotnet` console command now? I can give that a try.

**Comment by dsaf** [comment]
![image](https://user-images.githubusercontent.com/3845650/166226982-b76f2887-921f-43e8-8139-4a551fc926e5.png)

**Comment by MagicAndre1981** [comment]
[this is known for a while](https://github.com/xamarin/xamarin-android/issues/6652), the workaround is to use VS2022 17.**0**.x 0r VS2019 16.11.x

**Comment by dsaf** [comment]
> [this is known for a while](https://github.com/xamarin/xamarin-android/issues/6652), the workaround is to use VS2022 17.**0**.x 0r VS2019 16.11.x

Thank you, but I am not sure if that is the same issue. For example, why is the `Restart Adb Server` menu item disabled for me, yet there is no mention of this in #6652 ?

**Comment by MagicAndre1981** [comment]
I also see my Pixel phone in VS2022 17.2 preview 5, but deploy fails.

**Comment by dsaf** [comment]
> I also see my Pixel phone in VS2022 17.2 preview 5, but deploy fails.

For me no typical devices are present - and usually it should be at least the emulator but also a Samsung when connected.

**Comment by dellis1972** [comment]
This looks more like an IDE issue. It would probably be better if you reported the issue from within the IDE using the " Help > Send Feedback > Report a Problem" menu. 

The problem we have here is that the team who looks at this repo are not part of the IDE team.

**Comment by dsaf** [comment]
> This looks more like an IDE issue. It would probably be better if you reported the issue from within the IDE using the " Help > Send Feedback > Report a Problem" menu.

I have never seen a problem solved on those forums... Do you believe it's just something to do with my machine? I just want to understand if it's a local fluke, or if it affects many people.

> The problem we have here is that the team who looks at this repo are not part of the IDE team.

How do I initiate a Xamarin Android deployment from a command line? I can give working without an IDE a try, it's probably just the debugging that I would miss.

**Comment by dellis1972** [comment]
Those issues do get fixed. If you open one via the `Report a Problem` what happens is a bug will get created in the MS internal bug tracker. I can then put that infront of the Xamarin IDE extension team to take a look and investigate. If I point them to this bug the first thing they will say is "we don't track bugs on github" .


With regards to deploying from the command line, I can do one better for you. I documented a while back on how to get a Xamarin.Android project building and running (and debugging) in VSCode. See https://github.com/infinitespace-studios/xamarinvscode-sample/blob/main/Documentation/DebugXamarinAndroidWithVSCode.md for a how to and an example. It should work with .net 6 as well but it will need some changes to switch from `msbuild` to `dotnet`.  I might work on documenting that and update that example repo. 

But if you really want to test deployment from the command line use

```
msbuild Foo.Android.csproj /restore /t:Install /p:Configuration=Debug
```

This will do a basic build and install.

If you get any deployment issues for a device you will need to enable Fast Deployment diagnostic logging by adding
`<_FastDeploymentDiagnosticLogging>True</_FastDeploymentDiagnosticLogging>`  to your csproj. Or passing 
`/p:_FastDeploymentDiagnosticLogging=True` on the command line. Be warns this produces allot of data , in the IDE it can cause performance issues so do not switch it on for every project.

**Comment by dsaf** [comment]
I have been doing some experiments, and it looks like opening VS in the `Run as administrator` mode is strongly correlated with the functionality working. It might actually be the workaround - if it keeps working consistently. Though, I still like the idea of having an additional non-VS approach like the VS Code one below.

> Those issues do get fixed. If you open one via the `Report a Problem` what happens is a bug will get created in the MS internal bug tracker. I can then put that infront of the Xamarin IDE extension team to take a look and investigate. If I point them to this bug the first thing they will say is "we don't track bugs on github" .

Understood, thank you.

> With regards to deploying from the command line, I can do one better for you. I documented a while back on how to get a Xamarin.Android project building and running (and debugging) in VSCode. See https://github.com/infinitespace-studios/xamarinvscode-sample/blob/main/Documentation/DebugXamarinAndroidWithVSCode.md for a how to and an example. It should work with .net 6 as well but it will need some changes to switch from `msbuild` to `dotnet`. I might work on documenting that and update that example repo.

That is perfect, thank you! Regarding .NET 6 - I am waiting for MonoGame to be updated, which seems might happen very soon!

> But if you really want to test deployment from the command line use
> 
> ```
> msbuild Foo.Android.csproj /restore /t:Install /p:Configuration=Debug
> ```
> 
> This will do a basic build and install.

Also very useful!

> If you get any deployment issues for a device you will need to enable Fast Deployment diagnostic logging by adding `<_FastDeploymentDiagnosticLogging>True</_FastDeploymentDiagnosticLogging>` to your csproj. Or passing `/p:_FastDeploymentDiagnosticLogging=True` on the command line. Be warns this produces allot of data , in the IDE it can cause performance issues so do not switch it on for every project.

I was wondering about verbose logging, as otherwise there is few details to give to support - thank you!

I will close this ticket for now. Hopefully, anyone with the same issue searching will be able to find some helpful tips here.
Should the issue persist consistently, next step would be opening a VS support ticket using `Report a Problem` as recommended and linking it here.

**Comment by MagicAndre1981** [comment]
> I have been doing some experiments, and it looks like opening VS in the `Run as administrator` mode is strongly correlated with the functionality working. It might actually be the workaround - if it keeps working consistently. 

that also fixed my compile/deploy issue in 17.2 Preview 6 👍
